pkg/virt-handler/hotplug-disk/mount_test.go

```diff
@@ -61,1057 +64,936 @@ var (
 	orgFindMntByDevice   = findMntByDevice
 )
 
-var _ = Describe("HotplugVolume mount target records", func() {
+var _ = Describe("HotplugVolume", func() {
 	var (
-		m      *volumeMounter
-		err    error
-		vmi    *v1.VirtualMachineInstance
-		record *vmiMountTargetRecord
+		ctrl               *gomock.Controller
+		expectedCgroupRule *devices.Rule
+		cgroupManagerMock  *cgroup.MockManager
+		cgroupExpectedRuns int
+		cgroupCurrentRun   int
 	)
 
-	BeforeEach(func() {
-		tempDir, err = ioutil.TempDir("", "hotplug-volume-test")
-		Expect(err).ToNot(HaveOccurred())
-		vmi = v1.NewMinimalVMI("fake-vmi")
-		vmi.UID = "1234"
-
-		m = &volumeMounter{
-			podIsolationDetector: &mockIsolationDetector{},
-			mountRecords:         make(map[types.UID]*vmiMountTargetRecord),
-			mountStateDir:        tempDir,
-			hotplugDiskManager:   hotplugdisk.NewHotplugDiskWithOptions(tempDir),
+	expectCgroupRule := func(t devices.Type, major, minor int64, allow bool) {
+		expectedCgroupRule = &devices.Rule{
+			Type:  t,
+			Major: major,
+			Minor: minor,
+			Allow: allow,
 		}
-		record = &vmiMountTargetRecord{
-			MountTargetEntries: []vmiMountTargetEntry{
-				{
-					TargetFile: filepath.Join(tempDir, "test"),
-				},
-			},
-		}
-		err := m.setMountTargetRecord(vmi, record)
-		Expect(err).ToNot(HaveOccurred())
-		expectedBytes, err := json.Marshal(record)
-		Expect(err).ToNot(HaveOccurred())
-		bytes, err := ioutil.ReadFile(filepath.Join(tempDir, string(vmi.UID)))
-		Expect(err).ToNot(HaveOccurred())
-		Expect(bytes).To(Equal(expectedBytes))
-	})
+	}
 
-	AfterEach(func() {
-		os.RemoveAll(tempDir)
-		deviceBasePath = orgDeviceBasePath
-		statCommand = orgStatCommand
-		cgroupsBasePath = orgCgroupsBasePath
-		mknodCommand = orgMknodCommand
-	})
+	areRulesEqual := func(rule1, rule2 *devices.Rule) bool {
+		Expect(rule1).ToNot(BeNil())
+		Expect(rule2).ToNot(BeNil())
 
-	It("setMountTargetRecord should fail if vmi.UID is empty", func() {
-		vmi.UID = ""
-		record := &vmiMountTargetRecord{
-			MountTargetEntries: []vmiMountTargetEntry{
-				{
-					TargetFile: filepath.Join(tempDir, "test"),
-				},
-			},
-		}
-		err := m.setMountTargetRecord(vmi, record)
-		Expect(err).To(HaveOccurred())
-	})
+		return rule1.Type == rule2.Type &&
+			rule1.Major == rule2.Major &&
+			rule1.Minor == rule2.Minor &&
+			rule1.Allow == rule2.Allow
+	}
+
+	getCgroupManager = func(_ *v1.VirtualMachineInstance) (cgroup.Manager, error) {
+		return cgroupManagerMock, nil
+	}
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+
+		// cgroups mock setup
+		const defaultCgroupExpectedRuns = 2
+		cgroupExpectedRuns = defaultCgroupExpectedRuns
+		cgroupCurrentRun = 1
+		cgroupManagerMock = cgroup.NewMockManager(ctrl)
+		cgroupManagerMock.EXPECT().Set(gomock.Any()).Do(func(r *runc_configs.Resources) {
+			Expect(cgroupCurrentRun).To(BeNumerically("<=", cgroupExpectedRuns),
```

Comment:
Maybe this can be simplified somehow with:
```go
cgroupManagerMock.EXPECT().Set(gomock.Any()).Do(...).Times(defaultCgroupExpectedRuns)
```

Reply:
Yes, I would like tests to be able to modify the times they expect cgroups `Set()` to run dynamically.

For example, if one test expects 2 runs and another test expects 3 runs, I wanted it to be changed easily with `cgroupExpectedRuns = 3`. If this value is in an external variable that we compare to dynamically it can be changed, but doing `.Times(2)` is harder to change.

I ended up doing something like this:
```
mockSetFunc := func(r *runc_configs.Resources) {
     // implementation...
}

setExpectedCgroupRuns := func(runsExpected int) {
    groupManagerMock.EXPECT().Set(gomock.Any()).Do(mockSetFunc).Times(runsExpected)
}
```

This way `Times()` is utilized here but still can be set dynamically.
Thanks!